### DISCUSSION TITLE
Expected behaviour on minimum or maximum Y axis value change

### DISCUSSION THREAD

**Discussion by SSHshadow222** [patrykandpatrick/vico#674] [Questions] [answered]
I would like to know if the following behaviours are expected when changing the minimum or maximum values of the Y axis:

1. A transaction must occur in order for the changes to be visible.
1. Nothing changes if the series of the transaction remains unchanged.

Here's a code example containing a composable that you can easily integrate in a project to check the discussed behaviours (Please focus on the last two commented sections, and uncomment one at a time):

```kotlin
@Composable
private fun AppScreen(
    modifier: Modifier = Modifier
) {
    val modelProducer = remember { CartesianChartModelProducer.build() }
    var maxY by remember { mutableFloatStateOf(10f) }
    var minY by remember { mutableFloatStateOf(0f) }

    Column(
        verticalArrangement = Arrangement.spacedBy(16.dp),
        modifier = modifier
            .fillMaxSize()
            .padding(16.dp)
    ) {
        Column(
            verticalArrangement = Arrangement.spacedBy(8.dp),
            modifier = Modifier
                .weight(1f)
        ) {
            CartesianChartHost(
                chart = rememberCartesianChart(
                    rememberLineCartesianLayer(
                        axisValueOverrider = AxisValueOverrider.fixed(
                            minY = minY,
                            maxY = maxY
                        )
                    ),
                    startAxis = rememberStartAxis(
                        itemPlacer = AxisItemPlacer.Vertical.step({ 1f })
                    )
                ),
                modelProducer = modelProducer,
                modifier = Modifier
            )
            Text("minY = $minY")
            Text("maxY = $maxY")
        }
        HorizontalDivider()
        Row(
            horizontalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Column(
                verticalArrangement = Arrangement.spacedBy(4.dp),
                modifier = Modifier
                    .weight(1f)
            ) {
                Button(
                    onClick = { minY++ },
                    modifier = Modifier
                        .fillMaxWidth()
                ) {
                    Text("minY +")
                }
                Button(
                    onClick = { minY-- },
                    modifier = Modifier
                        .fillMaxWidth()
                ) {
                    Text("minY -")
                }
            }
            Column(
                verticalArrangement = Arrangement.spacedBy(4.dp),
                modifier = Modifier
                    .weight(1f)
            ) {
                Button(
                    onClick = { maxY++ },
                    modifier = Modifier
                        .fillMaxWidth()
                ) {
                    Text("maxY +")
                }
                Button(
                    onClick = { maxY-- },
                    modifier = Modifier
                        .fillMaxWidth()
                ) {
                    Text("maxY -")
                }
            }
        }
    }

//    // Doesn't have any effect when changing minY or maxY.
//
//    LaunchedEffect(minY, maxY) {
//        modelProducer.tryRunTransaction {
//            lineSeries {
//                series((1..10).toList())
//            }
//        }
//    }

//    // Changes the graph as expected.
//
//    LaunchedEffect(minY, maxY) {
//        modelProducer.tryRunTransaction {
//            lineSeries {
//                series((minY.toInt()..maxY.toInt()).toList())
//            }
//        }
//    }
}
```

Here's how the composable looks like:

![share_8798558713200311736](https://github.com/patrykandpatrick/vico/assets/113729758/f6e992e5-cf31-488b-8d68-eb967fd3b8c6)

As a side note, I would personally expect the minimum and maximum values of the Y axis to update the graph regardless of its data and to be independent of a transaction.

**Comment by patrickmichalik** [comment, marked as answer]
Hello! This is expected. When an `AxisValueOverrider` is applied to a `CartesianLayer`, it influences the `CartesianChart`’s `ChartValues` instance. `ChartValues` updates, in turn, occur only on `CartesianChartModelProducer.Transaction`s. This relates to performance optimization and, more importantly, animations.  
  
Still, you can dynamically update a `CartesianChart`’s _x_ and _y_ ranges without changing its series. To do so, use [extras](https://patrykandpatrick.com/vico/wiki/cartesian-charts/data#extras "https://patrykandpatrick.com/vico/wiki/cartesian-charts/data#extras") to store the ranges, create an `AxisValueOverrider` that reads the extras, and change the ranges by running a `CartesianChartModelProducer.Transaction` that updates the extras. As you’ve noted, `CartesianChartModelProducer.Transaction`s that change nothing are ignored. This, too, is an optimization measure. However, `CartesianChartModelProducer.Transaction`s that update just the extras are accepted. Generally, a `CartesianLayer`’s `AxisValueOverrider` should be set only once, but it can depend on extras whose values change over time, giving you full flexibility.

As previously mentioned, `AxisValueOverrider`s have a global effect. As it stands, there’s a single, explicit means of triggering updates to such global data: running `CartesianChartModelProducer.Transaction`s. These are the updates for which animations are triggered. The suggested behavior would require a second, less explicit update mechanism, which would have limited applicability—in cases where both the series and the _x_ and _y_ ranges are to be updated, extras would still be required for synchronization. The current system’s single-source-of-truth approach helps make it robust and universal. While extras may require a slight mindset shift, there’s a central, one-way channel for the kind of updates in question. This makes it clear—and gives you fine control over—how updates interact with each other, which can be useful for frequently updated charts. It’s thus our impression that in the grand scheme of things, the existing mechanism is optimal. Nonetheless, we’re certainly open to discussing this further.

If you need an example of the solution described in the second paragraph, or if you have any further questions, let me know. Cheers!
